File: acmicpc/acmicpc09012.cpp
#include <stdio.h>

using namespace std;
int main(){
    unsigned int n;
    scanf("%d", &n);

    for (int i=0; i < n; i++){
        char parenthesis[50];
        scanf("%s", parenthesis);

        int chk;
        chk = 0;

        for (int i=0; parenthesis[i]; i++){
            if (chk < 0){
                break;
            }
            if (parenthesis[i] == '('){
                chk += 1;
            } else {
                chk -= 1;
            }
        }
        if (chk == 0){
            printf("%s\n", "YES");
        } else {
            printf("%s\n", "NO");
        }
    }
}

File: acmicpc/acmicpc09012.py
import sys

n = int(sys.stdin.readline())
for _ in range(n):
    chk = 0
    parenthesis = sys.stdin.readline().strip()
    for s in parenthesis:
        if chk < 0:
            break
        if s == '(':
            chk += 1
        else:
            chk -= 1
    if chk == 0:
        print('YES')
    else:
        print('NO')
        